File: src/ip_filtering.py



#import libs
import argparse
import pandas as pd
import numpy as np
import vaex as vx


# Argument Parser
# -----------------
parser = argparse.ArgumentParser(description='Filter internal priming position')
parser.add_argument('bed_gtf_filtered', type=str, help='path of bed_gtf_filtered file')
parser.add_argument('ip_file', type=str, help='path of internal priming ref file')
parser.add_argument('output_path', type=str, help='path of output file for ip filtered reads')
parser.add_argument('output_bed_unique', type=str, help='path of output file for read mapped in unique transcript gene')
parser.add_argument('read_ids', type=str, help='path of output file for read ids')
args = parser.parse_args()


# File opening
# *********
#ip positions
ipdb = vx.read_csv(args.ip_file, sep="\t")
#reads
reads = vx.open(args.bed_gtf_filtered)


# filter ip position with exon only present among the reads (1)
ex_tofilter = (reads[['exon_id']].to_pandas_df().astype({'exon_id':'str'})).exon_id.unique().tolist()
ipdb ['exon_id'] = ipdb.exon_id.astype("str")
ipdb = ipdb[ipdb.exon_id.isin(ex_tofilter)]

if len(ipdb) != 0:
    # Annotate reads with internal priming coordinates limits(2)
    ip_plus = ipdb[ipdb.strand=="+"].groupby("transcript_name").agg({'start_ip':'max'})
    ip_plus.rename("start_ip","threshold_coord_ip")
    ip_neg = ipdb[ipdb.strand=="-"].groupby("transcript_name").agg({'end_ip':'min'})
    ip_neg.rename("end_ip","threshold_coord_ip")
    ipdb = vx.concat([ip_plus, ip_neg])


    # Delete all the reads at an upstream position of an internal priming interval (3)
    reads = reads.join(ipdb, on="transcript_name", how="left")
    reads['fidt'] = reads.frag_id + reads.transcript_name
    to_delete1 = reads[((reads.strand == "+") & (reads.end_rd<reads.threshold_coord_ip))][["fidt"]].to_pandas_df().fidt.tolist()
    to_delete2 = reads[((reads.strand == "-") & (reads.start_rd>reads.threshold_coord_ip))][["fidt"]].to_pandas_df().fidt.tolist()
    reads = reads[~reads.fidt.isin(to_delete1)]
    reads = reads[~reads.fidt.isin(to_delete2)]
    reads = reads.drop(["fidt","threshold_coord_ip"])



def drop_duplicates(df, columns=None):
    """Return a :class:`DataFrame` object with no duplicates in the given columns.
    .. warning:: The resulting dataframe will be in memory, use with caution.
    :param columns: Column or list of column to remove duplicates by, default to all columns.
    :return: :class:`DataFrame` object with duplicates filtered away.
    """
    if columns is None:
        columns = df.get_column_names()
    if type(columns) is str:
        columns = [columns]
    return df.groupby(columns, agg={'__hidden_count': vx.agg.count()}).drop('__hidden_count')


#get final reads ids to save for the bam filtering (4)
# read_ids = drop_duplicates(reads, columns="read_id")
read_ids = reads[['read_id','transcript_name']]


#select final columns to keep
reads = reads[["seqname", "start_rd", "end_rd", "strand", "bc", "umi","start","end","gene_name","gene_id","transcript_name","transcript_id","features","exon_number","rel_start","rel_end","salmon_rlp","rel_start_rd","rel_end_rd","dist_END"]]

# Filter reads associated to genes with unique transcripts (5)
reads_unique = reads[["gene_name","transcript_name"]].to_pandas_df().drop_duplicates().groupby("gene_name").count().reset_index(drop=False)
reads_unique = reads_unique[reads_unique.transcript_name == 1]
reads_unique = reads[reads.gene_name.isin(reads_unique.gene_name)]


#write files
reads.export_csv(args.output_path, sep="\t", index=False, doublequote=False, header=False)
reads_unique.export_csv(args.output_bed_unique, sep="\t", doublequote=False, index=False, header=False)
read_ids.export_csv(args.read_ids, sep="\t", index=False, doublequote=False, header=False)

File: src/ip_filtering.R


suppressMessages(suppressWarnings(library(argparse)))
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(stringr)))
suppressMessages(suppressWarnings(library(GenomicRanges)))


# Argument Parser
# +++++++++++++++
parser = ArgumentParser(description='Filtering of internal priming reads')
parser$add_argument('reads', type='character', help='path of bed_gtf file')
parser$add_argument('ip_file', type="character", help='path of internal priming ref file')
parser$add_argument('threshold_dist', type="character", help='distance of ip position from isoform end')
<<<<<<< HEAD
=======
parser$add_argument('encode_tab', type="character", help='encoding_table')
>>>>>>> 0c3fe80469feb05951ba0efe3f2fb8270c284a47
parser$add_argument('output_read_ids', type="character", help='path of output file for read ids mapped')
parser$add_argument('output_bed_unique', type="character", help='path of output file for read mapped in unique transcript gene')
parser$add_argument('output_ip', type="character", help='path of output ips')
parser$add_argument('output_path', type='character', help='path of output reads file')
args = parser$parse_args()

#params
THRESHOLD_IP_DIST = as.numeric(args$threshold_dist)

#0. Opening
<<<<<<< HEAD
#-----------
message("Opening...")
#reads
reads = readRDS(args$reads)
#ipdb
ipdb = fread(args$ip_file,col.names = c("seqnames.ip","start.ip","end.ip","c4","c5","strand.ip"), nThread=1)

if(nrow(ipdb)==0){

    #return void ipdb
    message("write ip sites mapped...")
    ipdb %>% fwrite(file=args$output_ip, sep="\t",row.names=F,col.names=F, nThread=2)

    #return reads
    #filtering
    reads = reads %>% 
     distinct(seqnames.rd,start.rd,end.rd,strand.rd,start.rdR,frag.id,start,end,gene_name,transcript_name,bulk_TPMperc,readID)

    #writing readID
    distinct(reads, readID) %>% fwrite(file=args$output_read_ids, sep="\t")
    reads = reads %>% dplyr::select(!readID)


    #4. Filter reads associated to genes with unique transcripts
    unique_genes = (reads %>% distinct(gene_name,transcript_name) %>% 
     group_by(gene_name) %>%
     summarise(count_tr = n()) %>%
     dplyr::filter(count_tr == 1))$gene_name
    reads_unique = reads %>% dplyr::filter(gene_name %in% unique_genes) %>% distinct()


    #5. Writing reads table
    message("writing...")
    reads = reads %>% dplyr::rename(dist_END="start.rdR")
    saveRDS(reads, file=args$output_path)
    #fwrite(reads, file=args$output_path, sep="\t", row.names=F, col.names=T, nThread=1)
    fwrite(reads_unique, file=args$output_bed_unique, sep="\t", row.names=F, col.names=F,  nThread=1)


}else {

    ipdb = ipdb %>% dplyr::select(!c(c4,c5))

    #Mapping
    #-------
    message("mapping...")
    annots = distinct(reads, seqnames.rd,start,end,startR,endR,strand,gene_name,transcript_name) ; annots
    hits = GenomicRanges::findOverlaps(
      GenomicRanges::makeGRangesFromDataFrame(annots, seqnames.field = "seqnames.rd", keep.extra.columns = T),
      GenomicRanges::makeGRangesFromDataFrame(ipdb, seqnames.field = "seqnames.ip", start.field = "start.ip",
                                              end.field = "end.ip", strand.field = "strand.ip")
    )
    message("internal priming processing...")
    ipdb = cbind(annots[hits@from,], ipdb[hits@to,]) %>%
      dplyr::filter(start.ip>start & end.ip<end) %>%
      #calculation of internal priming relative coordinates
      mutate(start.ipR = ifelse(strand=="-",endR-(end-start.ip),NA),
             end.ipR = ifelse(strand=="-",endR-(end-end.ip),NA),
             start.ipR = ifelse(strand=="+",endR-(end.ip-start),start.ipR),
             end.ipR = ifelse(strand=="+",endR-(start.ip-start),end.ipR)) %>%
      dplyr::filter(start.ipR>=THRESHOLD_IP_DIST) %>%
      #Get the closest IP out of threshold distance for each transcript
      group_by(transcript_name) %>%
      dplyr::filter(start.ipR == min(start.ipR)) %>%
      distinct(transcript_name,gene_name,start.ip,end.ip) %>%
      data.table(); ipdb

    #3. Write internal priming annotation sites
    message("write ip sites mapped...")
    ipdb %>% fwrite(file=args$output_ip, sep="\t",row.names=F,col.names=F, nThread=2)


    #Merging
    message("ip sites filtering...")
    trs_todel = (reads %>%
      left_join(ipdb) %>%
      #discard all fragment upstream of an IP position
      mutate(check=ifelse((!is.na(start.ip)) & strand=="+" & fg.end<start.ip,T,F),
             check=ifelse((!is.na(start.ip)) & strand=="-" & fg.start>end.ip,T,F),
             ftrs = paste0(transcript_name,"_",frag.id)) %>%
      dplyr::filter(check==T) %>%
      distinct(ftrs))$ftrs

    #filtering
    reads = reads %>%
      mutate(ftrs=paste0(transcript_name,"_",frag.id)) %>%
      dplyr::filter(!ftrs %in% trs_todel) %>%
      distinct(seqnames.rd,start.rd,end.rd,strand.rd,start.rdR,frag.id,start,end,gene_name,transcript_name,bulk_TPMperc,readID)

    #writing readID
    distinct(reads, readID) %>% fwrite(file=args$output_read_ids, sep="\t")
    reads = reads %>% dplyr::select(!readID)


    #4. Filter reads associated to genes with unique transcripts
    unique_genes = (reads %>% distinct(gene_name,transcript_name) %>%
                      group_by(gene_name) %>%
                      summarise(count_tr = n()) %>%
                      dplyr::filter(count_tr == 1))$gene_name
    reads_unique = reads %>% dplyr::filter(gene_name %in% unique_genes) %>% distinct()


    #5. Writing reads table
    message("writing...")
    reads = reads %>% dplyr::rename(dist_END="start.rdR")
    saveRDS(reads, file=args$output_path)
    fwrite(reads_unique, file=args$output_bed_unique, sep="\t", row.names=F, col.names=F,  nThread=1)

}
=======
#+++++++++++
message("Opening...")
#reads
reads = fread(args$reads, nThread =1)
#ipdb
ipdb = fread(args$ip_file,
             col.names = c("seqnames","start.ip","end.ip","c4","c5","strand.ip"), nThread=1)
#encode.tab
encode.tab = fread(args$encode_tab)

annot.extracted = distinct(reads, seqnames.rd,start,end,strand,transcript_name,gene_name,exon_id,exon_number,
                           start_rel,end_rel,tr_length)

#1. Processing
#+++++++++++++
message("Processing...")
if(length(ipdb)!=0){
  message("internal priming position associated to the chromosome detected...")
  
  #- annotate ip positions with isoforms infos
  hits.gr = findOverlaps(
    query = ipdb[,-c("c4","c5")] %>% 
      makeGRangesFromDataFrame(start.field = "start.ip", end.field = "end.ip", 
                               strand.field = "strand.ip", seqnames.field = "seqnames"),
    subject = annot.extracted %>% 
      makeGRangesFromDataFrame(keep.extra.columns = T, seqnames.field = "seqnames.rd")
  )
  
  #- merging
  ipdb = cbind(ipdb[hits.gr@from,-c("c4","c5")], annot.extracted[hits.gr@to,])
  ipdb = ipdb[,-c("seqnames.rd","strand")]
  
  #- filter ip positions
  ipdb = ipdb %>% filter(start.ip > min(start) & end.ip < max(end))
  
  #- calculate ip relative coords
  ipdb = lapply(c("+","-"), function(curr_strand){
    message(curr_strand)
    curr_ipdb = ipdb %>% filter(strand.ip == curr_strand)
    if(curr_strand=="+"){
      curr_ipdb = curr_ipdb %>% mutate(rel_start_ip = end_rel + (end-start.ip), rel_end_ip = end_rel + (end - end.ip))
      curr_ipdb$dist_END_ip = curr_ipdb$rel_end_ip
    }else{
      curr_ipdb = curr_ipdb %>% mutate(rel_start_ip = end_rel - (end - start.ip), rel_end_ip = end_rel - (end-end.ip))
      curr_ipdb$diste_END_ip = curr_ipdb$rel_start_ip
    }
    return(curr_ipdb)
  }) %>% rbindlist()
  
  #- Get the closest IP out of threshold distance for each transcript
  ipdb = ipdb %>%
    filter(dist_END_ip > THRESHOLD_IP_DIST) %>%
    group_by(transcript_name) %>%
    dplyr::filter(dist_END_ip == min(dist_END_ip)) %>%
    ungroup()
  
  #- Discard on each transcript, all the fragments with distEND < dist_END_ip
  frag_todel = left_join(reads,ipdb[,c("rel_start_ip","rel_end_ip","transcript_name","gene_name","dist_END_ip")],
                      multiple = "all") %>% na.omit() %>%
    dplyr::filter(dist_END_fg > dist_END_ip) %>%
    distinct(ft.encoded)
  reads = dplyr::filter(reads, !(ft.encoded %in% frag_todel$ft.encoded))
  
}else{
  
  message("no internal priming position associated to the chromosome detected...")
  
}
reads = reads[,-c("ft.encoded","dist_END_fg")]

#2. Get read_IDS encoded to filter
#+++++++++++++++++++++++++++++++++
encode.tab %>%
  dplyr::filter(read.id.encoded %in% reads$read.id.encoded) %>%
  distinct(read.id) %>%
  fwrite(args$output_read_ids,sep="\t",row.names = F,col.names = F)


#3. Write internal priming annotation sites
#++++++++++++++++++++++++++++++++++++++++++
ipdb %>% fwrite(file=args$output_ip, sep="\t",row.names=F,col.names=F, nThread=2)


#4. Filter reads associated to genes with unique transcripts
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
unique_genes = (reads %>% distinct(gene_name,transcript_name) %>% 
                  group_by(gene_name) %>%
                  summarise(count_tr = n()) %>%
                  filter(count_tr == 1))$gene_name
reads_unique = reads %>% filter(gene_name %in% unique_genes) %>% distinct()


#5. Writing reads table
fwrite(reads, file=args$output_path, sep="\t", row.names=F, col.names=T, nThread=1)
fwrite(reads_unique, file=args$output_bed_unique, sep="\t", row.names=F, col.names=T,  nThread=1)
>>>>>>> 0c3fe80469feb05951ba0efe3f2fb8270c284a47
